app/models/service/aggregation.rb

```diff
@@ -9,12 +9,121 @@ module Service::Aggregation
+  end
+
+  module ClassMethods
+    def aggregate_service_hardware_arel(aggregate_alias, hardware_column, options={})
+      virtual_column_name = "aggregate_all_vm_#{aggregate_alias}"
+
+      # if options[:mb_to_bytes]
+      #   aggregation_sql = (Hardware.arel_table[hardware_column] * 1.megabyte).sum
+      # else
+      #   aggregation_sql = Hardware.arel_table[hardware_column].sum
+      # end
+
+      # aggregation_sql = Hardware.arel_table[hardware_column]
+      # aggregation_sql = aggregation_sql * 1.megabyte if options[:mb_to_bytes]
+      # aggregation_sql = Arel::Nodes::NamedFunction.new('NUMERIC', [aggregation_sql.sum])
+
+      aggregation_sql = Hardware.arel_table[hardware_column].sum
+      aggregate_hardware_arel(virtual_column_name, aggregation_sql, options)
+    end
+
+    def aggregate_all_vms_disk_count_arel
+      virtual_column_name = "aggregate_all_vm_disk_count"
+      aggregation_sql = Disk.arel_table[:id].count
+      aggregate_hardware_arel(virtual_column_name, aggregation_sql, :include_disks => true)
+    end
+
+    def aggregate_all_vms_disk_space_allocated_arel
+      zero            = Arel::Nodes::SqlLiteral.new("0")
+      column_name     = "aggregate_all_vm_disk_space_allocated"
+      coalesce_values = [ Disk.arel_table[:size], zero ]
+      aggregation_sql = Arel::Nodes::NamedFunction.new('SUM',
+        [Arel::Nodes::NamedFunction.new('COALESCE', coalesce_values)]
+      )
+      aggregate_hardware_arel(column_name, aggregation_sql, :include_disks => true)
+    end
+
+    def aggregate_all_vms_disk_space_used_arel
+      zero            = Arel::Nodes::SqlLiteral.new("0")
+      column_name     = "aggregate_all_vm_disk_space_used"
+      coalesce_values = [ Disk.arel_table[:size_on_disk], Disk.arel_table[:size], zero ]
+      aggregation_sql = Arel::Nodes::NamedFunction.new('SUM',
+        [Arel::Nodes::NamedFunction.new('COALESCE', coalesce_values)]
+      )
+      aggregate_hardware_arel(column_name, aggregation_sql, :include_disks => true)
+    end
+
+    def aggregate_hardware_arel(virtual_column_name, aggregation_sql, options={})
+      lambda do |t|
+        subtree_services             = Arel::Table.new(:services)
+        subtree_services.table_alias = "#{virtual_column_name}_services"
+
+        service_resources   = ServiceResource.arel_table
+        vms                 = Vm.arel_table
+        hardwares           = Hardware.arel_table
+
+        vm_or_template_type = Arel::Nodes::SqlLiteral.new("'VmOrTemplate'")
+        ancestry_match      = Arel::Nodes::SqlLiteral.new("'/%'")
+        ancestry_ilike      = Arel::Nodes::NamedFunction.new("CONCAT", [t[:id], ancestry_match])
+        service_id_cast     = Arel::Nodes::NamedFunction.new("CAST", [t[:id].as("VARCHAR")])
+
+        where_clause   = t.grouping(
+          subtree_services[:id].eq(t[:id])
+           .or(subtree_services[:ancestry].matches(ancestry_ilike))
+        ).or(subtree_services[:ancestry].eq(service_id_cast))
```

Comment:
Looked into it and it seemed like it was calling `self.id`, meaning it was assuming that it was acting on an existing record downloaded from the DB, not the `id` column from the database.  I will double check this, since I would prefer to do this and re-use the scopes from the gem, but wrote it off as not possible.